proposals/mdm-baseline.md

```diff
@@ -0,0 +1,153 @@
+   enrollment profiles? See "macOS extensions" in MDM spec for details.
+
+### Problem: Issue an MDM command and monitor it’s status in fleet
+
+As described in
+[#Commands](./tools/mdm/apple/glossary-and-protocols.md#commands), sending and
+getting results from an MDM command is a multi stage process. 
+
+We should:
+
+1. Ensure all potential errors in each step are handled and _recorded_, so we
+   can report back to the user if there was an error.
+1. Ensure we have endpoints with desirable interfaces to run and get
+   information.
+
+#### Proposed solution
+
+1. Research and understand all the failure points when executing a command.
+1. Research and understand why issues like
+   [these](https://github.com/fleetdm/fleet/issues/8815) happen and figure out
+   if there's anything we can do.
+1. Ensure we're handling all errors as described in the MDM spec, including
+   `NotNow` commands.
+1. All MDM responses are XML, the server will be in charge of transforming those to JSON.
+    1. Discuss: this results in a heavier load in the server, should we
+       consider just delivering XML to the client to "distribute" this effort?
+1. Have an endpoint to retrieve all the commands that were issued to a host and
+   their status + results. 
+1. Have an endpoint that retrieves a single command and its status/results.
+    1. As a consequence of this, the UI (and possibly the CLI) will have to
+       "pull" for updates, we could consider upgrading this connection to a
+       WebSocket, leveraging the work already done for live queries.
+
+### Problem: Ingesting/storing data from MDM
+
+Some features require information that can only be accessed through MDM, for example:
+
+1. The recovery key for a host.
+1. The list of currently installed profiles.
+1. Security information.
+
+We need to stablish a pattern to ingest and keep this data in sync, similarly
+to how we do for data coming from `osquery`.
+
+Motivation:
+
+- https://github.com/fleetdm/fleet/issues/8519
+- https://github.com/fleetdm/fleet/issues/8360
+
+#### Proposed solution
+
+Implement a cron job that:
+
+1. Issues the necessary commands.
+1. Monitors their output.
+1. Ingests the data.
+
+Driving example: we need a list of the currently installed profiles in a host
+to check if its current status matches the desired status. To do this we need to:
+
+1. Issue a `ProfileList` command to all devices.
+1. Monitor its output until we get a response.
+1. In a table (real names TBD) `host_profiles` set `profiles`, `last_sync_at` and `sync_status`.
+
+
+#### TODO
+
+1. Timeout for command responses.
+1. Crash response, retries.?
+
+### Problem: Storing scripts, profile files and installers
+
+We currently store everything in MySQL, we should move all of this to a blob
+storage.
+
+We could start with S3, which is a standard API not tied to S3 to store the
+installers. We already have S3 logic to store installers and carves.
```

Comment:
Could we take a similar approach to carves and support both mysql and S3, with S3 as an option so that this is not a requirement? Or are we confident our users with MDM (I assume those assets are all MDM-related) will be ok with an S3-compatible dependency already?

Reply:
> Or are we confident our users with MDM (I assume those assets are all MDM-related) will be ok with an S3-compatible dependency already?

We should avoid making Fleet dependent on AWS or any other particular cloud platform.